src/D2L.CodeStyle.Analyzers/Immutability/TypeDeclarationImmutabilityAnalyzer.cs

```diff
@@ -57,26 +80,33 @@ INamedTypeSymbol typeSymbol
 			checker.CheckDeclaration( typeSymbol );
 		}
 
-		private static bool ShouldAnalyze(
-			INamedTypeSymbol analyzedType
+		private static void AnalyzeGenericMethodTypeArguments(
+			SyntaxNodeAnalysisContext ctx,
+			ImmutabilityContext immutabilityContext,
+			GenericNameSyntax syntax
 		) {
-			if( analyzedType.IsImplicitlyDeclared ) {
-				return false;
-			}
-
-			if( analyzedType.TypeKind == TypeKind.Interface ) {
-				return false;
+			SymbolInfo info = ctx.SemanticModel.GetSymbolInfo( syntax, ctx.CancellationToken );
+			if( !( info.Symbol is IMethodSymbol method ) ) {
+				return;
 			}
 
-			if( Attributes.Objects.Immutable.IsDefined( analyzedType ) ) {
-				return true;
-			}
+			int i = 0;
+			var paramArgPairs = method.TypeParameters.Zip( method.TypeArguments, ( p, a ) => (p, a, i++) );
+			foreach( var (parameter, argument, position) in paramArgPairs ) {
+				if( !Attributes.Objects.Immutable.IsDefined( parameter ) ) {
+					continue;
+				}
 
-			if( Attributes.Objects.ImmutableBaseClass.IsDefined( analyzedType ) ) {
-				return true;
+				if( !immutabilityContext.IsImmutable(
+					type: argument,
+					kind: ImmutableTypeKind.Total,
+					getLocation: () => syntax.TypeArgumentList.Arguments[position].GetLocation(),
+					out Diagnostic diagnostic
+				) ) {
+					// TODO: not necessarily a good diagnostic for this use-case

```

Comment:
This is maybe okay for type parameters (as demonstrated in the spec test) but won't be for actual types (nothing is being held by an immutable type)

Reply:
Yeah I think we'll need to do another pass generally...